plugins/woocommerce/src/Internal/Utilities/Types.php

```diff
@@ -0,0 +1,68 @@
+<?php
+
+namespace Automattic\WooCommerce\Internal\Utilities;
+
+use WP_Error;
+
+/**
+ * Utilities to help ensure type safety.
+ */
+class Types {
+	/**
+	 * Checks if the $thing is either the expected type, or a WP Error.
+	 *
+	 * If it passes the test then $thing will be returned unmodified, otherwise a WP_Error is returned. This is helpful
+	 * for ensuring return types are honored, but where the return value is filterable.
+	 *
+	 * Note that, in the context of REST API requests, the WP_Error may be exposed to the caller and therefore care
+	 * should be taken not to expose sensitive information.
+	 *
+	 * @since 9.1.0
+	 *
+	 * @param mixed           $thing        The return value to be assessed.
+	 * @param string          $desired_type What we expect the return type to be, if it is not a WP_Error.
+	 * @param string|null     $message      Optional message used to form a WP_Error if the type checks fail.
+	 * @param int|string|null $code         Optional code used to form a WP_Error if the type checks fail.
+	 * @param bool            $data         If $thing should be supplied to the WP_Error. Defaults to false.
+	 *
+	 * @return mixed|WP_Error
+	 */
+	public static function ensure_is_type_or_wp_error( $thing, string $desired_type, string $message = null, $code = null, bool $data = false ) {
+		// Is the $thing of the expected type?
+		if ( $thing instanceof $desired_type || is_wp_error( $thing ) ) {
```

Comment:
It looks like `instanceof` only works with objects. If our desired type is e.g. `array` or `integer` I don't think this will work.

Reply:
> I think we shouldn't assume it will only be used in the return statement at the end of a function, even though that's what's happening in this particular use case.

Agree.

> If our desired type is e.g. array or integer I don't think this will work.

That's right. As currently written, it's not intended to (but the method name probably doesn't make that very clear). 

My initial thought was that, if we're happy with the approach *in general,* we could add different methods like `Types::ensure_is_int_or_false`, to accommodate other common return types and unions. On which note:

> I'm on the fence about whether the or_wp_error part is too opinionated...

It seems like returning `WP_Error|WC_Data` is a very common pattern, in our REST API code in particular. Though, I guess we could have a far more generic and flexible approach that works something like this:

```pseudo
ensure_matches_type( $thing, 'builtin|udt|scalar' )
```

Parsing and evaluating the second param seems feasible, though there might be complexity we don't initially grasp.

Reply:
Perhaps also worth noting, the utility is in the internal namespace, so we have some freedom and flexibility in terms of being able to modify it in future (we don't need to care about back-compat for 3PD code).

Reply:
> Parsing and evaluating the second param seems feasible, though there might be complexity we don't initially grasp.

There could also be a middle ground. Something like:

```php
ensure_matches_type( $thing, $type, $fallback = null )
```

If `$fallback` is null, enforce the type no matter what. If `$fallback` is valid, use that if the type is wrong. Not sure it would make sense for `$fallback` to be any other arbitrary type, but certainly `WP_Error`, `false`, and maybe even throwing an exception would be good options. That would probably make for some messy logic though. Maybe `$fallback` is a callback, and the alternative return value can be defined there (including the particular error message within the WP_Error/Exception)?

Please excuse my stream of consciousness 😅

Reply:
That sounds interesting. Let me try something like that and see how it looks.

For the moment, I'll concentrate on objects rather than scalars (we can readily revise this in future, though, because as noted somewhere else the utility is internal).